src/controllers/forecast.mjs

```diff
@@ -1,23 +1,38 @@
 import { getForecastByCity } from "../services/forecast.mjs";
-import { getFileData } from "../utils/fileHelper.mjs";
-import { CITIES_FILE_PATH } from "../../config.mjs";
+import { getFileData, writeFile } from "../utils/fileHelper.mjs";
+import { CITIES_FILE_PATH, WEATHER_LOG_FILE_PATH } from "../../config.mjs";
+
+const checkUpdate = async (city) => {
+  const cityName = city.name.toLowerCase();
+  const weatherLog = await getFileData(WEATHER_LOG_FILE_PATH);
+  const cityWeather = weatherLog[cityName];
+  const twoHoursAgo = Date.now() - 2 * 60 * 60 * 1000;
+  const needUpdate =
+    !cityWeather || new Date(cityWeather.lastUpdate).getTime() < twoHoursAgo;
+
+  if (needUpdate) {
+    const forecast = await getForecastByCity(city);
+    weatherLog[cityName] = {
+      ...forecast,
+    };
+  }
+  await writeFile(WEATHER_LOG_FILE_PATH, weatherLog);
+};
 
 export const getForecastForCity = async (req, res) => {
   try {
     const cityName = req.params.cityName.toLowerCase();
-    const citiesData = await getFileData(CITIES_FILE_PATH);
-    const city = citiesData.find(
-      (city) => city.name.toLowerCase() === cityName
-    );
+    const cities = await getFileData(CITIES_FILE_PATH);
+    const city = cities.find((city) => city.name.toLowerCase() === cityName);
 
     if (!city) {
       return res.status(404).json({ message: "City not found" });
     }
-
-    const forecast = await getForecastByCity(city);
-    res.status(200).json(forecast);
-  } catch (error) {
-    console.error("Error getting forecast for city:", error);
+    await checkUpdate(city);
+    res.status(200).json((await getFileData(WEATHER_LOG_FILE_PATH))[cityName]);
```

Comment:
Due to main comment, we wouldn't need to read file one more time
Time saved, resources saved